File: github_code/avl-tree/java/31.java
package BinaryTree;

public class TreeAvl extends Tree 
{
	public Node Root;
	
	public int offSpring(Node node)
	{
		return (node.left != null) ? (node.right != null ? 2 : 1) : (node.right != null ? 1 : 0);
	}
	
	public Node rightRotate(Node nodo)
	{
		Node temp = nodo.left;
		Node temp2 = temp.right;
		
		temp.right = nodo;
		nodo.left = temp2;
		
		nodo.altulra = max(nodo.left.altulra, nodo.right.altulra) + 1;
		temp.altulra = max(temp.left.altulra, nodo.right.altulra) + 1;
		
		return temp;
		
	}
	
	public int getBalance(Node N) 
	{
	        if (N == null)
	            return 0;
	        else
	        	return this.height(N.left) - this.height(N.right);
	 }
	public Node leftRotate(Node nodo)
	{
		Node temp = nodo.right;
		Node Temp2 = temp.left;
		
		temp.left = nodo;
		nodo.right = Temp2;
		
		nodo.altulra = max(this.height(nodo.left), this.height(nodo.right)) + 1;
		temp.altulra = max(this.height(temp.left), this.height(nodo.right)) + 1;
		
		return temp;
	}
	
	 public int height(Node N)
	    {
	        if (N == null)
	             return 0;
	         return N.altulra;
	    }
	
	 public int max(int a, int b) 
	 {
		 return (a > b) ? a : b;
	 }
	 
	public Node insert(Node root, int toAdd)
	{
		if(root == null)
			return new Node(toAdd);
		
		if(toAdd < root.value)
			root.left = this.insert(root.left, toAdd);
		else if(toAdd > root.value)
			root.right = this.insert(root.right, toAdd);
		else
			return root;
		
		root.altulra = 1 + max(this.height(root.left), this.height(root.right));
		
		int balance = getBalance(root);
		
		 if (balance > 1 && toAdd < root.left.value)
			 return rightRotate(root);
		 
		 if (balance < -1 && toAdd > root.right.value)
			 return leftRotate(root);
		
		 if (balance > 1 && toAdd > root.left.value) 
		 {
			 root.left = leftRotate(root.left);
			 return rightRotate(root);
	     }
		 if (balance < -1 && toAdd < root.left.value) 
		 {
			 root.right = rightRotate(root.right);
			 return	leftRotate(root);
	     }
		 
		 return root;
	}
	
	public Node delete(Node root, int toDelete)
	{
		if(root == null)
			return root;
		if( toDelete < root.value)
			root.left =  this.delete(root.left, toDelete);
		else if(toDelete > root.value)
			root.right = this.delete(root.right, toDelete);
		else
		{
			 if (root.left == null)
	                return root.right;
	            else if (root.right == null)
	                return root.left;
	 
			  Node temp = minValueNode(root.right);
              root.value= temp.value;
	      
	            root.right = delete(temp, root.value);    
		}

		
        root.altulra = max(height(root.left), height(root.right)) + 1;
 
        int balance = getBalance(root);
 
        if (balance > 1 && getBalance(root.left) >= 0)
            return rightRotate(root);

        if (balance > 1 && getBalance(root.left) < 0)
        {
            root.left = leftRotate(root.left);
            return rightRotate(root);
        }
 
        if (balance < -1 && getBalance(root.right) <= 0)
            return leftRotate(root);
 
        if (balance < -1 && getBalance(root.right) > 0)
        {
            root.right = rightRotate(root.right);
            return leftRotate(root);
        }
 
		return root;
	}
	 Node minValueNode(Node node)
	    {
	        Node current = node;
	        while (current.left != null)
	           current = current.left;
	        return current;
	    }
	 void preOrder(Node node)
	    {
	        if (node != null)
	        {
	            System.out.print(node.value + " ");
	            preOrder(node.left);
	            preOrder(node.right);
	        }
	    }
	 
	    public static void main(String[] args)
	    {
	        TreeAvl tree = new TreeAvl();
	 
	        /* Constructing tree given in the above figure */
	        tree.root = tree.insert(tree.root, 9);
	        tree.root = tree.insert(tree.root, 5);
	        tree.root = tree.insert(tree.root, 10);
	        tree.root = tree.insert(tree.root, 0);
	        tree.root = tree.insert(tree.root, 6);
	        tree.root = tree.insert(tree.root, 11);
	        tree.root = tree.insert(tree.root, -1);
	        tree.root = tree.insert(tree.root, 1);
	        tree.root = tree.insert(tree.root, 2);
	 
	        /* The constructed AVL Tree would be
	           9
	          /  \
	         1    10
	        /  \    \
	        0    5    11
	        /    /  \
	        -1   2    6
	         */
	        System.out.println("Preorder traversal of "+
	                            "constructed tree is : ");
	        tree.preOrder(tree.root);
	 
	        tree.root = tree.delete(tree.root, 10);
	 
	        /* The AVL Tree after deletion of 10
	           1
	          /  \
	         0    9
	        /     / \
	        -1    5   11
	        /  \
	        2    6
	         */
	        System.out.println("");
	        System.out.println("Preorder traversal after "+
	                           "deletion of 10 :");
	        tree.preOrder(tree.root);
	    }
}

File: github_code/avl-tree/cpp/31.c
/*
 * avl.cpp
 */

#ifndef AVL_CPP
#define AVL_CPP



/* TODO
 * Goal: To find the balance factor of an AVL tree
 *       balance factor = height of right sub-tree - height of left sub-tree
 * Return: (int) balance factor
 */
template <typename T, typename K>
int AVL<T,K>::bfactor() const
{
    //write your codes here
	return this->is_empty() ? 0 : this->right_subtree()->height() - this->left_subtree()->height();
}


/* TODO
 * Goal: To rectify the height values of each node of of an AVL tree 
 */
template <typename T, typename K>
void AVL<T,K>::fix_height() const
{
    //write your codes here
	if (!this->is_empty()) {
		int left_height = this->left_subtree()->height();
		int right_height = this->right_subtree()->height();
		this->root->bt_height = 1 + max(left_height, right_height);
	}
}


/* TODO
 * Goal: To perform a single left (anti-clocwise) rotation of the root 
 */
template <typename T, typename K>
void AVL<T,K>::rotate_left() 
{
    //write your codes here
	typename AVL<T,K>::node* b = dynamic_cast<AVL*>(this->right_subtree())->root;
	this->root->right = b->left;
	b->left = dynamic_cast<AVL*>(this);
	fix_height();
	this->root = b;
	fix_height();
}


/* TODO
 * Goal: To perform right (clockwise) rotation of the root 
 */
template <typename T, typename K>
void AVL<T,K>::rotate_right()
{
     //write your codes here
	typename AVL<T,K>::node* b = dynamic_cast<AVL*>(this->left_subtree())->root;
	this->root->left = b->right;
	b->right = dynamic_cast<AVL*>(this);
	fix_height();
	this->root = b;
	fix_height();


/* TODO
 * Goal: To balance an AVL tree 
 */
template <typename T, typename K>
void AVL<T,K>::balance()
{
     //write your codes here
	if (this->is_empty()) {
		return;
	}
	fix_height();
	int balance_factor = this->bfactor();

	if (balance_factor == 2) {
		if (dynamic_cast<AVL*>(this->right_subtree())->bfactor() < 0) {
			dynamic_cast<AVL*>(this->right_subtree())->rotate_right();
		}
		return rotate_left();
	}
	else if (balance_factor == -2) {
		if (dynamic_cast<AVL*>(this->left_subtree())->bfactor() > 0) {
			dynamic_cast<AVL*>(this->left_subtree())->rotate_left();
		}
		return rotate_right();
	}
}


/* TODO
 * Goal: To insert an item x with key k to AVL tree
 */
template <typename T, typename K>
void AVL<T,K>::insert(const T& x, const K& k)
{
     //write your codes here
	if (this->is_empty()) {
		this->root = new typename AVL<T,K>::node(x,k);
		this->root->left = new AVL<T,K>;
		this->root->right = new AVL<T,K>;
	}

	else if (k < this->root->key)
		this->left_subtree()->insert(x, k);

	else if (k > this->root->key)
		this->right_subtree()->insert(x,k);

	balance();
}



/* TODO
 * Goal: To remove an item with key k in AVL tree 
 */
template <typename T, typename K>
void AVL<T,K>::remove(const K& k)
{
     //write your codes here
	if (this->is_empty()) {
		return;
	}

	if (k < this->root->key) {
		this->left_subtree()->remove(k);
	}

	else if (k > this->root->key) {
		this->right_subtree()->remove(k);
	}
	else {
		AVL<T,K>* left_avl = dynamic_cast<AVL*>(this->left_subtree());
		AVL<T,K>* right_avl = dynamic_cast<AVL*>(this->right_subtree());
		if (!left_avl->is_empty() && !right_avl->is_empty()) {
			this->root->value = dynamic_cast<AVL*>(right_avl->find_min())->root->value;
			this->root->key = dynamic_cast<AVL*>(right_avl->find_min())->root->key;
			right_avl->remove(this->root->key);
		}
		else {
			typename AVL<T,K>::node* node_to_remove = this->root;
			if (left_avl->is_empty()) {
				this->root = right_avl->root;
			}
			else {
				this->root = left_avl->root;
			}
			right_avl->root = left_avl->root = NULL;
			delete node_to_remove;
		}
	}

	balance();
}



#endif /* AVL_CPP */
